src/ocrd/resource_manager.py

```
@@ -196,30 +198,28 @@ def list_installed(self, executable: str = None):
                         'size': res_size
                     }
                 else:
-                    resdict = self.add_to_user_database(this_executable, res_filename, resource_type=res_type)
+                    resdict = self.add_to_user_database(this_executable, res_name, res_filename, resource_type=res_type)
                 # resdict['path'] = str(res_filename)
                 reslist.append(resdict)
             ret.append((this_executable, reslist))
         self.save_user_list()
         return ret
 
-    def add_to_user_database(self, executable, res_filename, url=None, resource_type='file'):
+    def add_to_user_database(self, executable, res_name, res_filename, url=None, resource_type='file'):
```

Why do we need the additional parameter `res_name` if we can already get it from `res_filename`? If it is for flexibility reasons, I would suggest having `res_name` with a default value of `Path(res_filename).name` unless there is a reason not to do that.

---

For example, in the Tesseract case, the [pretrained script models](https://github.com/tesseract-ocr/tessdata_best/tree/main/script) are usually referenced as e.g. `script/Latin.traineddata`. But we could of course flatten this in our tessdata installations. (And we could also try to get rid of the `configs/` resource...)

Another example is the new ocrd_pagetopdf's `data/AletheiaSans.ttf`, but that could just be a silent default again.

All other module resources I can find are in the top level. So perhaps we should reconsider the recursive dive in `list_all_resources` at the module location – if we make that flat as well, we can switch back to the `res_filename.name` paradigm (and simple strings instead of location-name tuples)...

---

> But we could of course flatten this in our tessdata installations. (And we could also try to get rid of the configs/ resource...)

Not sure how complicated that would be to change. 

> if we make that flat as well, we can switch back to the res_filename.name paradigm (and simple strings instead of location-name tuples)...

Not necessary to switch back, I guess; Nested resources should still be handled properly by `core` if there was no spec guiding how to store the resources.

---

> > But we could of course flatten this in our tessdata installations. (And we could also try to get rid of the configs/ resource...)
> 
> Not sure how complicated that would be to change.

We moved away from installing under tessdata/script/ a while ago. Some users might have to update their installations, but it would not be hard to enforce (code-wise). 

And the `configs` download was just a convenience for the standalone CLI – we could just make that part of the installation procedure, without a resmgr entry.

> > if we make that flat as well, we can switch back to the res_filename.name paradigm (and simple strings instead of location-name tuples)...
> 
> Not necessary to switch back, I guess; Nested resources should still be handled properly by `core` if there was no spec guiding how to store the resources.

What is astonishing is that the [current spec](https://github.com/OCR-D/spec/blob/master/ocrd_tool.md#file-parameters) states pretty much the converse of what was implemented:
- module resources must be directly under the top-level module directory
- all other resources can be arbitrary relative paths

So if we want to stick with that, we would need to keep my location-name tuple structure, but change the iterators in `list_all_resources` to support recursion. (But it might be hard to get the level right: take Calamari for example again – you must not descend into `fraktur_19th_century/?.ckpt/`. So the actual maxlevel seems to then depend on the processor, which ocrd_utils does not know anything about.)

I am inclined to rather change the spec: flat top-level everywhere.

---

> I am inclined to rather change the spec: flat top-level everywhere.

Agree!

---

So in accordance with https://github.com/OCR-D/spec/pull/263 I changed back the retval of `list_all_resources` to list of str, and disabled recursive moduled discovery – all locations but CWD are flat. I also added some basic tests for `list_all_resources`, which I found completely lacking.